### PR TITLE
Complete: Shared user-editable span recovery helpers (8.1.2)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Status: COMPLETE
- Implements the plan for 8.1.2: shared user-editable span recovery helpers for macro-heavy `rstest` code paths. This turns the prior planning material into an implemented foundation with a pure policy in `whitaker-common`, a thin `rustc_span::Span` adapter in `whitaker`, and an adopter in `function_attrs_follow_docs` demonstrating the approach. Includes tests, docs, and execution-plan updates.

## Rationale
- Move from planning material toward a concrete implementation that can be exercised by a real adopter and validated with unit and behavioural tests.
- Enforce a clean policy boundary: pure policy in `whitaker-common`, adapter in `whitaker`, and a single adopter to demonstrate correctness and reduce crate-local logic in macro-heavy paths.
- Provide gating, tests, and documentation to prevent scope creep and ensure future updates stay aligned with 8.1.2 goals.

## Changes
- Add new span-recovery module and tests:
  - common/src/rstest/span.rs (new)
  - common/src/rstest/mod.rs (exports updated)
  - common/src/lib.rs (exported SpanRecoveryFrame, UserEditableSpan, recover_user_editable_span)
  - common/src/rstest/tests.rs (unit tests for recovery policy)
  - common/tests/features/rstest_span_recovery.feature (Behaviour-driven tests)
  - common/tests/rstest_span_recovery_behaviour.rs (BDD-based tests)
- Implemented rustc-span adapter and helpers:
  - src/hir.rs (span_recovery_frames and recover_user_editable_hir_span added)
  - src/lib.rs (re-exports for adapter behind dylint-driver feature)
- Adopter integration:
  - crates/function_attrs_follow_docs/src/driver.rs updated to use the shared recovery adapter.
  - crates/function_attrs_follow_docs/src/tests/order_detection.rs updated for new surface usage.
- Documentation and exec plan:
  - docs/execplans/8-1-2-shared-user-editable-span-recovery-helpers.md added (ExecPlan with status COMPLETE)
  - docs/lints-for-rstest-fixtures-and-test-hygiene.md updated with 8.1.2 implementation decisions
  - docs/roadmap.md updated to mark 8.1.2 as done
- Misc:
  - Updated internal exports and surface in src/lib.rs to expose recovery helpers via the existing `whitaker` facade when the dylint-driver feature is enabled.

## Goals and implementation outline (updated)
- Pure policy surface in `whitaker-common` for determining if a frame in a span chain is user-editable and whether recovery is possible.
- Thin `rustc_span::Span` adapter in `whitaker` (via `src/hir.rs`) to walk the source-callsite chain and produce an ordered frame list consumable by the policy.
- At least one real consumer (demonstrated by `function_attrs_follow_docs`) adopting the shared helper to reduce crate-local logic and properly handle macro-generated glue.
- Documentation updates reflecting decisions and final design when implemented.

## Implementation stages (summary)
- Stage B: Define the contract with tests first (failing tests define the contract).
- Stage C: Implement the pure shared policy (unit tests and docs).
- Stage D: Implement the `rustc_span::Span` adapter (surface and re-exports behind `dylint-driver`).
- Stage E: Adopt the helper in one consumer (e.g., update `function_attrs_follow_docs` and regression tests).
- Stage F: Update design/docs and roadmap with implementation decisions.
- Stage G: Mark roadmap item 8.1.2 as done after gates pass.
- Stage H/I: Run full gates (`make fmt`, `make markdownlint`, `make nixie`, `make check-fmt`, `make lint`, `make test`).

## Validation and evidence (planning gates)
- The plan outlines unit tests and BDD-style tests demonstrating the contract (direct, recovered, macro-only, and first-match behavior).
- End-of-turn gates include format, lint, and test runs consistent with 8.1.2 gates. ExecPlan notes reflect that gates pass at turn end.

## Documentation impact
- ExecPlan added at `docs/execplans/8-1-2-shared-user-editable-span-recovery-helpers.md` and marked COMPLETE.
- Design decisions captured in `docs/lints-for-rstest-fixtures-and-test-hygiene.md`.
- Roadmap updated to reflect item 8.1.2 as DONE once implementation and gates are satisfied.

## Notes for reviewers
- This PR transitions from planning to implementation for 8.1.2 and introduces new public surface on `whitaker-common` and `whitaker` behind a feature gate. One adopter demonstrates viability and helps validate policy correctness and integration points.
- No user-facing runtime API changes beyond the new shared surface and its adoption in a consumer example.

## Task link
- Task: https://www.devboxer.com/task/ffb7bb88-d24d-410e-958f-8ca741eea97f

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

## Summary by Sourcery

Introduce shared rstest span recovery helpers and adopt them in an existing rstest-related lint path to avoid diagnostics on macro-only glue.

New Features:
- Add a pure rstest span recovery policy in whitaker-common that classifies ordered span frames as direct, recovered, or macro-only user-editable spans.
- Expose rustc_span::Span-based span recovery and frame-construction helpers in whitaker for use by compiler-integrated lints.
- Adopt the shared user-editable span recovery helper in the function_attrs_follow_docs lint to drive attribute ordering and item-bound checks from recovered user spans.

Enhancements:
- Export the new span recovery types and helpers through existing whitaker-common and whitaker facades behind the dylint-driver feature.

Documentation:
- Document the 8.1.2 implementation decisions and span recovery behavior, add an execplan for the shared helpers, and update the roadmap to mark 8.1.2 complete.

Tests:
- Add unit tests and rstest-bdd behavioral scenarios for the shared span recovery policy, and extend function_attrs_follow_docs tests to cover recovered span ordering and macro-only attribute filtering.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[common/src/rstest/tests.rs](https://github.com/leynos/whitaker/pull/197/files/5f37701b5ae00e9be9832421f93daa7c2ccb7efb#diff-7e24eab32ceb067c70cc83895d4c913f175c42585f3cb156827a1ef7ad5a6247)

Comment on file

~~~rust
fn treats_all_expansion_frames_as_macro_only() {
    let first = source_span(4, 1, 4);
    let second = source_span(5, 1, 6);

    assert_eq!(
        recover_user_editable_span(&[
            SpanRecoveryFrame::new(first, true),
            SpanRecoveryFrame::new(second, true),
        ]),
        UserEditableSpan::MacroOnly
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71836/delta?repo-id=1205552&review-id=197&biomarker=Code+Duplication&filename=common%2Fsrc%2Frstest%2Ftests.rs&method=)
The module contains 2 functions with similar structure: recovers_macro_frame_to_first_user_span,treats_all_expansion_frames_as_macro_only

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for the following concern and provide an AI coding agent prompt for the fix:

### ❌ Failed checks (3 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing | ⚠️ Warning | Public functions `span_recovery_frames` and `recover_user_editable_hir_span` lack unit test coverage, and two review suggestions remain unaddressed in panic messaging and test case implementation. | Add unit tests for the exported functions covering span scenarios, update panic message with line context, and implement the missing test case for `None` item span condition. |
| User-Facing Documentation | ⚠️ Warning | The PR implements user-facing behaviour changes to `function_attrs_follow_docs` (macro-aware span recovery and macro-only attribute dropping) that are documented in design docs but absent from the users guide. | Update docs/users-guide.md `function_attrs_follow_docs` section with macro-aware span recovery behaviour, macro-only attribute handling, and rstest-specific examples. |
| Developer Documentation | ⚠️ Warning | The PR introduces new internal APIs and architectural patterns but fails to update the repository's primary developer documentation (docs/developers-guide.md) with guidance on span recovery helpers and the policy-vs-adapter pattern. | Add a new 'Shared Span Recovery Helpers' section to docs/developers-guide.md documenting SpanRecoveryFrame, UserEditableSpan, and recover\_user\_editable\_hir\_span() with usage examples and architectural pattern explanation. |

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- In `recover_user_editable_hir_span`, you build an intermediate `Vec<SpanRecoveryFrame<Span>>` only to pass `as_slice` into `recover_user_editable_span`; consider iterating the callsite chain directly and feeding the policy without allocation to avoid per-call heap traffic on hot paths.
- The new `attribute_within_item` logic returns `false` whenever `item_user_editable_span` is `None`, even if the raw item span is non-dummy; consider falling back to `raw_item_span` in that case so you don’t accidentally drop all attributes for items whose spans cannot be recovered.

## Individual Comments

### Comment 1
<location path="crates/function_attrs_follow_docs/src/tests/order_detection.rs" line_range="99-108" />
<code_context>
+    assert_eq!(recover_user_editable_span(&frames), expected);
+}
+
+#[rstest]
+fn keeps_direct_user_editable_span() {
+    let span = source_span(1, 1, 8);
</code_context>
<issue_to_address>
**suggestion (testing):** Add tests for `attribute_within_item` covering dummy item spans, `None` item spans, and out-of-bounds recovered attribute spans.

To fully exercise the updated `attribute_within_item` logic, please add tests for: (1) a dummy `raw_item_span` that always returns `true`; (2) `item_span: None` returning `false` even when the attribute is within the raw item span; and (3) an attribute span extending beyond the item span to confirm it’s rejected. These will cover the remaining branches without going through the full lint path.
</issue_to_address>

### Comment 2
<location path="docs/execplans/8-1-2-shared-user-editable-span-recovery-helpers.md" line_range="3" />
<code_context>
+# Add shared user-editable span recovery helpers for macro-heavy `rstest` code paths
+
+This ExecPlan is a living document. The sections `Constraints`, `Tolerances`,
+`Risks`, `Progress`, `Surprises & Discoveries`, `Decision Log`, and
+`Outcomes & Retrospective` must be kept up to date as work proceeds.
</code_context>
<issue_to_address>
**suggestion (review_instructions):** ExecPlan is used as an acronym without being defined on first use.

Consider expanding this on first use, for example: `This Execution Plan (ExecPlan) is a living document.` Subsequent occurrences can then safely use the shortened form.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing | ⚠️ Warning | Session globals wrapper missing from expanded\_span() test helper in src/hir.rs and out-of-bounds rejection test absent from order\_detection.rs. | Wrap expanded\_span() body with create\_default\_session\_globals\_then() and add raw\_item\_span\_rejects\_out\_of\_bounds\_attribute\_when\_item\_recovery\_fails() test case. |
| User-Facing Documentation | ⚠️ Warning | User-facing documentation in `docs/users-guide.md` has not been updated to address the review comment at lines 309–313 regarding misleading scope description of the lint. | Update `docs/users-guide.md` section "The lint checks:" (around line 311) to consolidate bullets and reframe `#[rstest]` and `#[test]` as macro-heavy examples, clarifying the lint's scope encompasses all function types with outer attributes. |

File: src/hir.rs [hunk 1]
```diff
@@ -240,3 +283,120 @@ fn module_has_harness_descriptor<'tcx>(
...
+            let (span, expected_frames, expected_recovered_span) = build_span_case(case);
+
+            assert_eq!(span_recovery_frames(span), expected_frames);
+            assert_eq!(
+                recover_user_editable_hir_span(span),
+                expected_recovered_span
+            );
+        });
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/hir.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: src/hir.rs [hunk 2]
```diff
@@ -53,6 +53,49 @@ pub fn has_test_like_hir_attributes(
...
+        Some(span)
+    })
+}
+
+#[must_use]
+pub fn span_recovery_frames(span: Span) -> Vec<SpanRecoveryFrame<Span>> {
+    walk_span_chain(span)
+        .map(|current| SpanRecoveryFrame::new(current, current.from_expansion()))
+        .collect()
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/hir.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai pause
